### PR TITLE
i18n: Add `--lang` option to `make_rst.py` to write translated headings

### PR DESCRIPTION
**Pull request by akien-mga**
Add a few missing headings to the translation templates.

Note: Adds a dependency on [`polib`](https://pypi.org/project/polib/), but only for using the new option.
It can be used like this:
```
cd doc
make rst LANGARG=zh_CN
```
or manually by passing `--lang zh_CN` or similar to the `make_rst.py` call.

### CODE REVIEW

File: doc/tools/make_rst.py [hunk 1]
```diff
@@ -1261,7 +1299,9 @@ def make_method_signature(
     return ret_type, out
 
 
-def make_heading(title, underline):  # type: (str, str) -> str
+def make_heading(title, underline, l10n=True):  # type: (str, str, bool) -> str
+    if l10n:
+        title = headings_l10n.get(title, title)
     return title + "\n" + (underline * len(title)) + "\n\n"
```
Comment (timothyqiu) [inline doc/tools/make_rst.py:1305]: reStructuredText requires the underlines' length to be visually the same as the title at least.

While after l10n, `title` may contain wide characters, so it should be `len(title) * 2` here.

File: doc/tools/make_rst.py [hunk 2]
```diff
@@ -376,6 +399,21 @@ def main():  # type: () -> None
     )
     args = parser.parse_args()
 
+    # Retrieve heading translations for the given language.
+    if not args.dry_run:
+        lang_file = os.path.join(
+            os.path.dirname(os.path.realpath(__file__)), "..", "translations", "{}.po".format(args.lang)
+        )
+        if os.path.exists(lang_file):
+            import polib
```
Comment (timothyqiu) [inline doc/tools/make_rst.py:408]: Not a requirement, but we can use something like this to produce a nicer error message when `polib` is missing.

```python
try:
    import polib
except ImportError:
    print("Section header localization requires `polib`.")
    exit(1)
```

File: doc/tools/make_rst.py (hunk 1 above)
Comment (akien-mga) [reply doc/tools/make_rst.py:1305]: Hm, but Sphinx also throws warnings if the underlines' length is not *exactly* the same as the text :|
Can we figure out what the exact correct length would be?

Comment (akien-mga) [reply doc/tools/make_rst.py:1305]: Might be worth looking up what test Sphinx itself performs to see if we can replicate it. Alternatively, maybe there's a lint hint that we can set in generated files to ignore the warning.

Comment (timothyqiu) [reply doc/tools/make_rst.py:1305]: Ah, it uses something like this: https://github.com/docutils-mirror/docutils/blob/e88c5fb08d5cdfa8b4ac1020dd6f7177778d5990/docutils/utils/__init__.py#L618-L642

Unicode combining characters are also taken into account 😲

Comment (akien-mga) [reply doc/tools/make_rst.py:1305]: Yeah I was about to link that :D

But I also found that I'm mistaken, it doesn't complain if the underline is too long, only if it's too short. So your solution to multiply by 2 is fine :+1:

Comment (akien-mga) [reply doc/tools/make_rst.py:1311]: Should be fixed now.

File: doc/tools/make_rst.py [hunk 3]
```diff
@@ -19,6 +19,28 @@
 MARKUP_ALLOWED_PRECEDENT = " -:/'\"<([{"
 MARKUP_ALLOWED_SUBSEQUENT = " -.,:;!?\\/'\")]}>"
 
+# Used to translate the section headings when required with --lang argument.
+# The HEADINGS list should be synced with what we actually write with `make_heading`,
+# and also hardcoded in `doc/translations/extract.py`.
+HEADINGS = [
```
Comment (akien-mga) [inline doc/tools/make_rst.py:25]: Note to self: When cherry-picking to `3.x`, the Constructors and Operators strings are not relevant.

File: doc/tools/make_rst.py [hunk 4]
```diff
@@ -1261,7 +1303,11 @@ def make_method_signature(
     return ret_type, out
 
 
-def make_heading(title, underline):  # type: (str, str) -> str
+def make_heading(title, underline, l10n=True):  # type: (str, str, bool) -> str
+    if l10n:
+        if title in headings_l10n:
+            title = headings_l10n.get(title)
+            underline *= 2  # Double length to handle wide chars.
```
Comment (bruvzg) [inline doc/tools/make_rst.py:1310]: Maybe it should be something like `(underline * len(title.decode('utf-8')))` to get the actual number of Unicode characters (not sure if it's what it wants).

Comment (timothyqiu) [reply doc/tools/make_rst.py:1310]: `title` is a `str` in Python 3 I believe, so no need/way to decode 😛

**Review by timothyqiu — APPROVED** [review]
Looks good to me :)

File: doc/tools/make_rst.py (hunk 4 above)
Comment (akien-mga) [reply doc/tools/make_rst.py:1310]: It's actually the visual width in monospace columns that we want, but that's tricky to get, see https://github.com/godotengine/godot/pull/56127/files/09563dc665dbda970b84f5c63955155b8f8857e6#diff-8cfde96f29054d38628a87af04b23d77280174c3d31df0c38a23cbd34bce960a

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.

**Pull request merged**